### PR TITLE
Code refactor and fix condition bug

### PR DESCRIPTION
**Pull request by GermanAizek**
Code improvements both in readability and excess code parts

### CODE REVIEW

File: client/CServerHandler.cpp [hunk 1]
```diff
@@ -489,7 +489,7 @@ void CServerHandler::sendMessage(const std::string & txt) const
 		std::string connectedId, playerColorId;
 		readed >> connectedId;
 		readed >> playerColorId;
-		if(connectedId.length(), playerColorId.length()) // BUG https://bugs.vcmi.eu/view.php?id=3144
+		if(connectedId.length() < playerColorId.length())
```
Comment (kambala-decapitator) [inline client/CServerHandler.cpp:492]: reading the following code, I don't find the fix correct. It looks like the intention was to check that strings are not empty.

File: lib/CBonusTypeHandler.cpp [hunk 2]
```diff
@@ -94,7 +94,9 @@ std::string MacroString::build(const GetValue & getValue) const
 CBonusType::CBonusType()
 {
 	hidden = true;
-	icon = nameTemplate = descriptionTemplate = "";
+	icon.clear();
+    nameTemplate.clear();
+    descriptionTemplate.clear();
```
Comment (kambala-decapitator) [inline lib/CBonusTypeHandler.cpp:99]: we use tabs for indentation
```suggestion
	nameTemplate.clear();
	descriptionTemplate.clear();
```

File: lib/CGameState.cpp [hunk 3]
```diff
@@ -726,7 +726,7 @@ void CGameState::init(const IMapService * mapService, StartInfo * si, bool allow
 	logGlobal->info("\tUsing random seed: %d", si->seedToBeUsed);
 	getRandomGenerator().setSeed(si->seedToBeUsed);
 	scenarioOps = CMemorySerializer::deepCopy(*si).release();
-	initialOpts = CMemorySerializer::deepCopy(*si).release();
+	initialOpts = scenarioOps;
```
Comment (kambala-decapitator) [inline lib/CGameState.cpp:729]: are you absolutely sure that this is correct? To me it looks that the original intention was to keep these entities completely separate.

File: lib/mapObjects/CGTownInstance.cpp [hunk 4]
```diff
@@ -46,7 +46,7 @@ void CCreGenAsCastleInfo::serializeJson(JsonSerializeFormat & handler)
 
 	if(!handler.saving)
 	{
-		asCastle = (instanceId != "");
+		asCastle = (!instanceId.empty());
```
Comment (kambala-decapitator) [inline lib/mapObjects/CGTownInstance.cpp:49]:
```suggestion
		asCastle = !instanceId.empty();
```

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]
Please split the first commit into two: one that actually fixes the reported bug, the other with code improvements.

Btw have you used some static analyser to find those issues?

File: lib/CGameState.cpp (hunk 3 above)
Comment (Nordsoft91) [reply lib/CGameState.cpp:729]: This is incorrect and will break restart scenario functionality

Comment (Nordsoft91) [reply lib/CGameState.cpp:729]: This is incorrect and will break restart scenario functionality

File: client/CServerHandler.cpp (hunk 1 above)
Comment (Nordsoft91) [reply client/CServerHandler.cpp:492]: Exactly, this command assigns specified player color to connectionId. It doesn't make any sense to compare string length. Passing connectionId="5" and playerColorId="1" would be absolutely valid but fails with proposed change

File: lib/CGameState.cpp (hunk 3 above)
Comment (Nordsoft91) [reply lib/CGameState.cpp:729]: this change is incorrect and will break restart scenario functionality

File: client/CServerHandler.cpp (hunk 1 above)
Comment (nullkiller) [reply client/CServerHandler.cpp:492]: Still the syntax is weird. Cpp ref does not say it is legal to put , inside if and what does it mean

Comment (nullkiller) [reply client/CServerHandler.cpp:492]: Looks like it is comma operator and according to it the first statement result will be just discurded

Comment (kambala-decapitator) [reply client/CServerHandler.cpp:492]: > Looks like it is comma operator and according to it the first statement result will be just discurded

yes, you're absolutely correct

Comment (GermanAizek) [reply client/CServerHandler.cpp:492]: @kambala-decapitator OK, then I will remove from this PR changes affected `client/CServerHandler.cpp`

File: lib/mapObjects/CGTownInstance.cpp (hunk 4 above)
Comment (GermanAizek) [reply lib/mapObjects/CGTownInstance.cpp]: @kambala-decapitator fix it here https://github.com/vcmi/vcmi/pull/1111/commits/2c01e9680c3778d5f4a2732eb6f7d01c42fc2501

File: lib/CGameState.cpp (hunk 3 above)
Comment (GermanAizek) [reply lib/CGameState.cpp]: @Nordsoft91 fix it here https://github.com/vcmi/vcmi/pull/1111/commits/2c01e9680c3778d5f4a2732eb6f7d01c42fc2501

File: lib/CBonusTypeHandler.cpp (hunk 2 above)
Comment (GermanAizek) [reply lib/CBonusTypeHandler.cpp]: @kambala-decapitator fix it https://github.com/vcmi/vcmi/pull/1111/commits/2c01e9680c3778d5f4a2732eb6f7d01c42fc2501

File: client/windows/CAdvmapInterface.cpp [hunk 5]
```diff
@@ -573,7 +573,7 @@ CAdvMapInt::CAdvMapInt():
 	strongInterest = true; // handle all mouse move events to prevent dead mouse move space in fullscreen mode
 	townList.onSelect = std::bind(&CAdvMapInt::selectionChanged,this);
 	bg = BitmapHandler::loadBitmap(ADVOPT.mainGraphic);
-	if (ADVOPT.worldViewGraphic != "")
+	if (!ADVOPT.worldViewGraphic.empty())
```
Comment (Nordsoft91) [inline client/windows/CAdvmapInterface.cpp:576]: We don’t use spaces after for if while pls fix here and in other placs too

**Review by nullkiller — APPROVED** [review]

File: client/windows/CAdvmapInterface.cpp (hunk 5 above)
Comment (GermanAizek) [reply client/windows/CAdvmapInterface.cpp]: @Nordsoft91 fix it here https://github.com/vcmi/vcmi/pull/1111/commits/7fdad4e0f6edc1fae14b1008360f9bc2a9d885ef

**Review by Nordsoft91 — APPROVED** [review]

**Pull request merged**